src/pedlar.ts

```diff
@@ -16,23 +22,54 @@ export class Pedlar {
    * return another function that when invoked will clean up the
    * original effect.  The returned function will automatically be
    * invoked when the side effect is destroyed.
-   * @returns The ID of this side effect, that can be used to
-   * individually clean it up by calling the `destroy() method.
+   * @returns An object containing:
+   *   - The ID of this side effect that can be used to individually
+   *    clean it up by calling the `destroy()` method
+   *   - A `performAgain()` function that can be used to perform the side
+   *     effect again, after automatically calling the destroyer
    */
-  public perform(effect: Function): string {
+  public perform<D extends any[]>(
+    effect: Function,
+    dependencies?: D,
+  ): PerformanceResult<D> {
     let destroyer = effect()
+    let id = generateId()
+
+    if (destroyer !== undefined && destroyer !== null) {
+      validateDestroyer(destroyer)
+      this.destroyers[id] = destroyer
+    }
+
+    return {
+      pedlar: this,
+      id,
+      dependencies,
+      effect,
+      destroyer,
+      performAgain(currentDependencies?: D) {
+        let hasChanged = compare(this.dependencies, currentDependencies)
+        this.dependencies = currentDependencies
+        if (!hasChanged) return
 
-    if (destroyer === undefined || destroyer === null) return
-    if (typeof destroyer !== 'function') {
-      throw new Error(
-        'Effect must return undefined, null, or a function; found: ' +
-          typeof destroyer,
-      )
+        let destroyer = this.pedlar.destroyers[this.id]
+        if (destroyer) destroyer()
+
+        let newDestroyer = this.effect()
+        if (newDestroyer === undefined || newDestroyer === null) {
```

Comment:
This code allows for the effect to return a destroyer even though it did not originally return one.  Because the effect cannot change from one run to the next, this case is not possible.  All the destroyer validation can happen in the original `perform()` call.